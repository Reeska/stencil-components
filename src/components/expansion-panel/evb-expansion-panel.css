
:host {
  display: block;
  margin-top: 8px;
}

.body {
  margin: 1rem 0;
}

.title {
  font-size: 1.25em;
  cursor: pointer;
  display: block;
  outline: none;
  padding: .5rem 0;
  user-select: none;
}

.title.right {
  text-align: right;
}

.title.center {
  text-align: center;
}

.title svg {
  display: inline-flex;
  fill: currentColor;
  height: 1rem;
  margin-left: .5rem;
  vertical-align: middle;
  width: 1rem;
}
