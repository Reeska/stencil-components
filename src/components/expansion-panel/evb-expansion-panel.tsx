import { Component, ComponentInterface, Event, EventEmitter, Method, Prop } from '@stencil/core';

@Component({
  tag: 'evb-expansion-panel',
  shadow: true,
  styleUrl: 'expansion-panel.css'
})
export class EvbExpansionPanel implements ComponentInterface {
  /**
   * The textual title of the panel
   */
  @Prop() text: string;
  /**
   * Where to align the title
   */
  @Prop() justify?: 'left' | 'right' | 'center' = 'left';
  /**
   * Whether the panel is open, eg: shows it's content
   */
  @Prop({ mutable: true }) open = false;

  /**
   * Emits when the panel is closed
   */
  @Event() closed: EventEmitter<void>;
  /**
   * Emits when the panel is opened
   */
  @Event() opened: EventEmitter<void>;

  constructor() { }

  ngOnInit() { }

  @Method()
  toggle(force?: boolean) {
    this.open = force !== undefined ? force : !this.open;
    this.open ? this.opened.emit() : this.closed.emit();
  }

  render() {
    const classes = ['title', this.justify];
    return ([
      <div class={classes.join(' ')} onClick={() => this.toggle()}>{this.text}</div>,
      this.open ? <slot></slot> : ''
    ]);
  }

}
