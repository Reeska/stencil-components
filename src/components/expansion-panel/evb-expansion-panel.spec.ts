import { EvbExpansionPanel } from './evb-expansion-panel';

describe('EvbExpansionPanel', () => {

  let component: EvbExpansionPanel;

  beforeEach(() => component = new EvbExpansionPanel());

  it('should create', () => {
    expect(component).toBeTruthy();
    expect(component.render()).toMatchSnapshot();
  });

  it('should render text', () => {
    component.text = 'Hello world';
    component.open = true;
    expect(component.render()).toMatchSnapshot();
  });

  it('should emit when toggled', () => {
    const openedSpy = jest.fn();
    const closedSpy = jest.fn();

    component.opened = { emit: openedSpy };
    component.closed = { emit: closedSpy };

    expect(component.open).toBeFalsy();
    expect(openedSpy).not.toHaveBeenCalled();
    expect(closedSpy).not.toHaveBeenCalled();

    component.toggle(); // open
    expect(component.open).toBeTruthy();
    expect(openedSpy).toHaveBeenCalledTimes(1);
    expect(closedSpy).not.toHaveBeenCalled();

    component.toggle(); // close
    expect(component.open).toBeFalsy();
    expect(openedSpy).toHaveBeenCalledTimes(1);
    expect(closedSpy).toHaveBeenCalledTimes(1);
  });
});
