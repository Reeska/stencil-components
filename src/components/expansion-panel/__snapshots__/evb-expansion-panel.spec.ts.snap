// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EvbExpansionPanel should create 1`] = `
Array [
  Object {
    "elm": undefined,
    "ishost": false,
    "vattrs": Object {
      "class": "title left",
      "onClick": [Function],
    },
    "vchildren": Array [
      Object {
        "vtext": "",
      },
    ],
    "vkey": undefined,
    "vname": undefined,
    "vtag": "div",
    "vtext": undefined,
  },
  "",
]
`;

exports[`EvbExpansionPanel should render text 1`] = `
Array [
  Object {
    "elm": undefined,
    "ishost": false,
    "vattrs": Object {
      "class": "title left",
      "onClick": [Function],
    },
    "vchildren": Array [
      Object {
        "vtext": "Hello world",
      },
    ],
    "vkey": undefined,
    "vname": undefined,
    "vtag": "div",
    "vtext": undefined,
  },
  Object {
    "elm": undefined,
    "ishost": false,
    "vattrs": null,
    "vchildren": null,
    "vkey": undefined,
    "vname": undefined,
    "vtag": "slot",
    "vtext": undefined,
  },
]
`;
